### PR TITLE
shellcheck: lint bash files

### PR DESCRIPTION
**Pull request by oliver-sanders**
Equivalent of https://github.com/cylc/cylc/pull/3068

* Run strict checks over all functional scripts.
* Run lenient checks over test scripts.
* Fix existing code to make the new checks pass

Some of the scripts modified here have little to no testing so take care reviewing:
* `bin/rose-make-docs`
* `etc/*`
* `sbin/*`

### CODE REVIEW

**Comment by oliver-sanders** [comment]
I think the remaining failure will be fixed in #2318

**Comment by oliver-sanders** [comment]
Rebased, deconflicted.

File: .travis/shellchecker [hunk 1]
```diff
@@ -0,0 +1,92 @@
...
+
+    # run shellcheck
+    shellcheck "$@" "${files[@]}"
+}
+
+# default configuration for linting cylc
+default () {
+    # run a strict check on all "functional" scripts
+    main bin etc lib/bash sbin usr t/lib/bash \
+        --exclude etc/rose-bash-completion \
```
Comment (oliver-sanders) [inline .travis/shellchecker]: I did have a go at fixing this file but it proved to be flaky. We will need to re-write it with the retirement of `rose suite-run` anyway.

**Comment by oliver-sanders** [comment]
BTW, `shellcheck` has rather nice documentation outlining each issue which may help with reviewing e.g. 
[SC2086](https://github.com/koalaman/shellcheck/wiki/SC2086)

File: .travis/shellchecker (hunk 1 above)
Comment (oliver-sanders) [reply .travis/shellchecker]: I had a couple of goes at this file but I kept breaking the tests. This script will have to be re-written (or at least heavily edited) soon.

**Comment by oliver-sanders** [comment]
Addressed all feedback except `etc/rose-bash-completion` which I'm keen to leave out as I don't seem to be able to make it pass without breaking it and the script will need to be re-written soon with `rose suite-run` migration.

**Comment by oliver-sanders** [comment]
Rebased.

File: .travis/shellchecker [hunk 2]
```diff
@@ -0,0 +1,91 @@
...
+    readarray -t files < <(for x in "${files[@]}"; do echo "$x"; done | sort -u)
+
+    # run shellcheck
+    shellcheck "$@" "${files[@]}"
+}
+
+# default configuration for linting cylc
+default () {
+    # run a strict check on all "functional" scripts
+    main bin etc lib/bash sbin usr t/lib/bash \
```
[the remark this answers is not in the document]
Comment (sadielbartholomew) [reply .travis/shellchecker]: Also I believe there are only two Bourne shell files under ``etc``, so we could replace ``etc ... --exclude ...`` with just the remaining ``etc/deploy-docs``, but I appreciate that the current form might be more maintainable & clear for the future when we will remove the exclusion (& maybe add other bash files there).

File: bin/rose-make-docs [hunk 3]
```diff
@@ -203,11 +201,12 @@ version_file () {
...
     echo '{'
     # scrape filesystem for list of rose versions which have built docs
-    VERSIONS=( $(cd "${DOCS_DIR}"; echo $DOC_VERSIONS) )
+    mapfile -t VERSIONS < <(cd "${DOCS_DIR}"; echo "${DOC_VERSIONS[@]}")
     for version in "${VERSIONS[@]}"; do
         # scrape filesystem to get list of formats this version is available in
-        formats=( $(cd "${DOCS_DIR}/${version}"; echo $DOC_FORMATS) )
-        list="    \"${version}\": $(json_list ${formats[@]})"
+        mapfile -t formats \
+            < <(cd "${DOCS_DIR}/${version}"; echo "${DOC_FORMATS[@]}")
```
Comment (sadielbartholomew) [inline bin/rose-make-docs]: When I ran ``rose make-docs`` they built cleanly & the built docs looked fine, but something about this line caused unhappiness:
```bash
Build finished. The HTML pages are in <...>/rose.git/doc/2019.01.0/html.
make: Leaving directory `<...>/rose.git/sphinx'
<...>/rose.git/bin/rose-make-docs: line 207: cd: <...>/rose.git/doc/ACKNOWLEDGEMENT.md CONTRIBUTING.md COPYING README.md bin demo etc lib licences pytest.ini rose-version sbin sphinx t tester_temp.py tox.ini usr venv: No such file or directory
```

File: .travis/shellchecker (hunk 2 above)
[the remark this answers is not in the document]
Comment (oliver-sanders) [reply .travis/shellchecker]: Added all scripts to shellcheck tests.

File: bin/rose-make-docs [hunk 4]
```diff
@@ -190,31 +186,26 @@ venv-destroy () {
...
-            echo "${list}"
-        else
-            echo "${list},"
-        fi
+        ret+="\n    \"$version\": ["
+        for format_dir in "${version_dir}/"!(doctrees|*.html); do
+            format=$(basename "$format_dir")
+            ret+="$format, "
+        done
+        ret="${ret:0:-2}],"
```
Comment (oliver-sanders) [inline bin/rose-make-docs]: FYI `ret=${ret:0:-1}` is equivalent to backspace.

Comment (sadielbartholomew) [reply bin/rose-make-docs]: open_mouth I was aware that this kind of thing (extracting a sub-string by indexing ultimately, I guess) could be done in Python, but not in bash. Nice!

File: .travis/now [hunk 5]
```diff
@@ -76,6 +76,7 @@ _install_cylc () {
 }
 
 _install_fcm () {
+    # shellcheck disable=SC1004
```
[the remark this answers is not in the document]
Comment (oliver-sanders) [reply .travis/now]: Note to all, this was due to `.travis/shellchecker` being run by a much older version of shellcheck.

Even though this is a Travis script I'll put in a fix as this is the `shellcheck` version we have been shipped with our OS.

**Pull request merged**